File: 173_BinarySearchTreeIterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
##24.37%
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self._array = list()
        self._currIndex = 0
        if not root:
            return
        stack = list()
        curr = root
        while stack or curr:
            if curr:
                if curr.left:
                    stack.append(curr)
                    curr = curr.left
                else:
                    self._array.append(curr.val)
                    curr = curr.right
            else:
                curr = stack.pop()
                self._array.append(curr.val)
                curr = curr.right

    def hasNext(self):
        """
        :rtype: bool
        """
        return self._currIndex < len(self._array)
        

    def next(self):
        """
        :rtype: int
        """
        self._currIndex += 1
        return self._array[self._currIndex - 1]
        

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())
###Clearly a better solution
class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.stack = list()
        self.pushAll(root)

    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return self.stack

    # @return an integer, the next smallest number
    def next(self):
        tmpNode = self.stack.pop()
        self.pushAll(tmpNode.right)
        return tmpNode.val
        
    def pushAll(self, node):
        while node is not None:
            self.stack.append(node)
            node = node.left

File: 173_BinarySearchTreeIterator.cpp
//37.12%
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode *root):currIndex(0) {
		//We did a inorder tracing here and obtain 'inorderArray'
		if(!root)
			return;
        stack<TreeNode *> s;
		//s.push(root);
		TreeNode *curr(root);
		while(!s.empty() || curr){
			if(curr){
				if(curr -> left){
					s.push(curr);
					curr = curr -> left;
				}else{
					inorderArray.push_back(curr -> val);
					curr = curr -> right;
				}
			}else{
				curr = s.top();
				s.pop();
				inorderArray.push_back(curr -> val);
				curr = curr -> right;
			}
		}
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return currIndex < inorderArray.size();
    }

    /** @return the next smallest number */
    int next() {
        return inorderArray[currIndex++];
    }
private:
	vector<int> inorderArray;
	int currIndex;
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */
 
//Clearly a better solution from LeetCode 
class BSTIterator {
    stack<TreeNode *> myStack;
public:
    BSTIterator(TreeNode *root) {
        pushAll(root);
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !myStack.empty();
    }

    /** @return the next smallest number */
    int next() {
        TreeNode *tmpNode = myStack.top();
        myStack.pop();
        pushAll(tmpNode->right);
        return tmpNode->val;
    }

private:
    void pushAll(TreeNode *node) {
        for (; node != NULL; myStack.push(node), node = node->left);
    }
};